cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/ReportGenerationSpark.java

```diff
@@ -0,0 +1,379 @@
+      }
+      if (limit > MAX_LIMIT) {
+        responder.sendError(400, "limit must cannot be larger than " + MAX_LIMIT);
+        return;
+      }
+      Location reportBaseDir = getDatasetBaseLocation(ReportGenerationApp.REPORT_FILESET).append(reportId);
+      if (!reportBaseDir.exists()) {
+        responder.sendError(404, String.format("Report with id %s does not exist.", reportId));
+        return;
+      }
+      // Get the status of the report and only COMPLETED report can be read
+      ReportStatus status = getReportStatus(reportBaseDir);
+      if (!ReportStatus.COMPLETED.equals(status)) {
+        responder.sendError(400, String.format("Report with id %s with status %s cannot be read.", reportId, status));
+        return;
+      }
+      List<String> reportRecords = new ArrayList<>();
+      long lineCount = 0;
+      Location reportDir = reportBaseDir.append(REPORT_DIR);
+      Location reportFile = null;
+      // TODO: assume only one report file for now
+      for (Location file : reportDir.list()) {
+        if (file.getName().endsWith(".json")) {
+          reportFile = file;
+          break;
+        }
+      }
+      if (reportFile == null) {
+        responder.sendError(500, "No files found for report " + reportId);
+        return;
+      }
+      // Read the report file and add lines starting from the position of offset to the result until the result reaches
+      // the limit
+      try (BufferedReader br = new BufferedReader(new InputStreamReader(reportFile.getInputStream()))) {
+        String line;
+        while ((line = br.readLine()) != null) {
+          // skip lines before the offset
+          if (lineCount++ < offset) {
+            continue;
+          }
+          if (reportRecords.size() == limit) {
+            break;
+          }
+          reportRecords.add(line);
+        }
+      }
+      // Get the total number of records from the COUNT file
+      String total =
+        new String(ByteStreams.toByteArray(reportBaseDir.append(COUNT_FILE).getInputStream()), Charsets.UTF_8);
+      responder.sendJson(200, new ReportContent(offset, limit, Integer.parseInt(total), reportRecords));
+    }
+
+    @POST
+    @Path("/reports")
+    public void executeReportGeneration(HttpServiceRequest request, HttpServiceResponder responder)
+      throws IOException {
+      String requestJson = Charsets.UTF_8.decode(request.getContent()).toString();
+      LOG.debug("Received report generation request {}", requestJson);
+      ReportGenerationRequest reportRequest;
+      try {
+        reportRequest = decodeRequestBody(requestJson, REPORT_GENERATION_REQUEST_TYPE);
+        reportRequest.validate();
+      } catch (IllegalArgumentException e) {
+        responder.sendError(400, e.getMessage());
+        return;
+      }
+
+      String reportId = ReportIds.generate().toString();
+      Location reportBaseDir = getDatasetBaseLocation(ReportGenerationApp.REPORT_FILESET).append(reportId);
+      reportBaseDir.mkdirs();
```

Comment:
`mkdirs` could fail, in which case it returns false, you need to check the return value and return error message in that case.